File: Version-JavaScript/models/Supplier.js
import Person from './Person.js';

// Class to Manage Supplier who is above all a Person
class Supplier extends Person {
    constructor(name, contactInfo, gender) {
        super(name, contactInfo, gender);
        this.productsSupplied = []; // List of supplied products
    }

	// Add a supplier to inventory of supermaket
    addSupplier(supermarket) {
        supermarket.suppliers.push(this);
        console.log(`Supplier "${this.name}" added to supermarket "${supermarket.name}".`);
    }

	// Method to get Supplier informations
	getSupplierInfos() {
		console.log(`
		Supplier Infos:
		-----------------
		
		ID: ${this.id}
		Name: ${this.name}
		Contact Info: ${this.contactInfo}
		Sexe: ${this.gender}
		Products Supplied: ${this.productsSupplied.length}
		`);
	};
	
	// Method to delete a supplier from a suppermaket
    deleteSupplier(supermarket) {
        supermarket.suppliers = supermarket.suppliers.filter(supplier => supplier.id !== this.id);
        console.log(`Supplier "${this.name}" removed from supermarket "${supermarket.name}".`);
    };
    
    // Method to supply a product depending of the productID and the quantity
	supplyProduct(supermarket, productID, quantity) {
		if(!supermarket, !productID || !quantity) {
			throw new Error("All required informations should be provided !!!");
		}
		
		const product = supermarket.products.find(p => p.id === productID);

        if (!product) throw new Error("Product not found!");
        product.stock += quantity;
		
		this.productsSupplied.push(product);
		
		// Return success message after all
		console.log(`${quantity} Products of ID "${productID}" has been successful supplied !`);
	};
}
export default Supplier;

// Test
const supplier = new Supplier("SupplierA", "supplierA@example.com", "Female");

// supplier.getSupplierInfos();

File: Version-Python/models/supplier.py
from .person import Person

class Supplier(Person):
    def __init__(self, name, contact_info, gender):
        super().__init__(name, contact_info, gender)
        self.products_supplied = []

    def add_supplier(self, supermarket):
        supermarket.suppliers.append(self)
        print(f'Supplier "{self.name}" added to supermarket "{supermarket.name}".')

    def get_supplier_infos(self):
        print(f"""
        Supplier Infos:
        -----------------
        
        ID: {self.id}
        Name: {self.name}
        Contact Info: {self.contact_info}
        Sexe: {self.gender}
        Products Supplied: {len(self.products_supplied)}
        """)

    def delete_supplier(self, supermarket):
        supermarket.suppliers = [s for s in supermarket.suppliers if s.id != self.id]
        print(f'Supplier "{self.name}" removed from supermarket "{supermarket.name}".')

    def supply_product(self, supermarket, product_id, quantity):
        if not all([supermarket, product_id, quantity]):
            raise ValueError("All required information should be provided!")

        product = next((p for p in supermarket.products if p.id == product_id), None)
        if not product:
            raise ValueError("Product not found!")

        product.quantity += quantity
        self.products_supplied.append(product)
        print(f'{quantity} Products of ID "{product_id}" has been successful supplied!')
